ext/webgpu/01_webgpu.js

```diff
@@ -0,0 +1,7055 @@
+ * @param {InnerGPUDevice} device
+ * @returns {GPUQueue}
+ */
+function createGPUQueue(label, device) {
+  /** @type {GPUQueue} */
+  const queue = webidl.createBranded(GPUQueue);
+  queue[_label] = label;
+  queue[_device] = device;
+  return queue;
+}
+
+class GPUQueue {
+  /** @type {InnerGPUDevice} */
+  [_device];
+
+  constructor() {
+    webidl.illegalConstructor();
+  }
+
+  /**
+   * @param {GPUCommandBuffer[]} commandBuffers
+   */
+  submit(commandBuffers) {
+    webidl.assertBranded(this, GPUQueuePrototype);
+    const prefix = "Failed to execute 'submit' on 'GPUQueue'";
+    webidl.requiredArguments(arguments.length, 1, {
+      prefix,
+    });
+    commandBuffers = webidl.converters["sequence<GPUCommandBuffer>"](
+      commandBuffers,
+      prefix,
+      "Argument 1",
+    );
+    const device = assertDevice(this, prefix, "this");
+    const commandBufferRids = ArrayPrototypeMap(
+      commandBuffers,
+      (buffer, i) => {
+        const context = `command buffer ${i + 1}`;
+        const rid = assertResource(buffer, prefix, context);
+        assertDeviceMatch(device, buffer, {
+          prefix,
+          selfContext: "this",
+          resourceContext: context,
+        });
+        return rid;
+      },
+    );
+    const { err } = ops.op_webgpu_queue_submit(device.rid, commandBufferRids);
+    for (let i = 0; i < commandBuffers.length; ++i) {
+      commandBuffers[i][_rid] = undefined;
+    }
+    device.pushError(err);
+  }
+
+  onSubmittedWorkDone() {
+    webidl.assertBranded(this, GPUQueuePrototype);
+    return PromiseResolve();
+  }
+
+  /**
+   * @param {GPUBuffer} buffer
+   * @param {number} bufferOffset
+   * @param {BufferSource} data
+   * @param {number} [dataOffset]
+   * @param {number} [size]
+   */
+  writeBuffer(buffer, bufferOffset, data, dataOffset = 0, size) {
+    webidl.assertBranded(this, GPUQueuePrototype);
+    const prefix = "Failed to execute 'writeBuffer' on 'GPUQueue'";
+    webidl.requiredArguments(arguments.length, 3, prefix);
+    buffer = webidl.converters["GPUBuffer"](buffer, prefix, "Argument 1");
+    bufferOffset = webidl.converters["GPUSize64"](
+      bufferOffset,
+      prefix,
+      "Argument 2",
+    );
+    data = webidl.converters.BufferSource(data, prefix, "Argument 3");
+    dataOffset = webidl.converters["GPUSize64"](
+      dataOffset,
+      prefix,
+      "Argument 4",
+    );
+    size = size === undefined
+      ? undefined
+      : webidl.converters.GPUSize64(size, prefix, "Argument 5");
+    const device = assertDevice(this, prefix, "this");
+    const bufferRid = assertResource(buffer, prefix, "Argument 1");
+    assertDeviceMatch(device, buffer, {
+      prefix,
+      selfContext: "this",
+      resourceContext: "Argument 1",
+    });
+    const { err } = ops.op_webgpu_write_buffer(
+      device.rid,
+      bufferRid,
+      bufferOffset,
+      dataOffset,
+      size,
+      new Uint8Array(ArrayBufferIsView(data) ? data.buffer : data),
```

Comment:
`TypedArrayPrototypeGetBuffer` should be used for `TypedArray`, and `DataViewPrototypeGetBuffer` should be used for `DataView`.

https://github.com/denoland/deno/blob/3f96e5a29a88afafcef0f17458b2800b2db316ee/ext/web/08_text_encoding.js#L112-L124